### ISSUE TITLE
VBA-Dependency overview

### ISSUE THREAD

**Issue by joshkreud** [rubberduck-vba/Rubberduck#3004]
Rubberduck Currently Parses the Code and knows wich procedure depends on what other Procedure and Module. 

So Would it be possible to create an excel or csv exporter that creates some output like this?:

![image](https://cloud.githubusercontent.com/assets/10895412/25803735/97d631dc-33f8-11e7-915b-85908acd7ad3.png)

Where Depends[n] shows Module.Procedure / Class.Procedure

This could help creating some sort of a Code Map. 

Greetings 
Josh

**Comment by retailcoder** [comment]
I believe it's very well possible to write VBA code that does exactly that, using the Rubberduck COM API: if you have early-bound unit tests in your project, you already have a reference to the Rubberduck TLB - one of the types exposed is `ParserState`; the feature is experimental and will very likely change in possibly breaking ways in future versions, but you can definitely use it to iterate user declarations and their respective `IdentifierReference` objects, each of which contains a reference to its parent/scoping `Declaration` (that would be the calling procedure).

Something like (phone post, untested):

````vb
Sub Test
    With New Rubberduck.ParserState
        .Initialize Application.VBE
        .Parse 'blocking call... for now
        Dim item As Variant, userDeclaration As Declaration
        Dim callScope As Declaration
        Dim refItem As Variant, callSite As IdentifierReference
        
        For Each item In .UserDeclarations
            Set userDeclaration = item 'explicit cast is required... for now
            For Each refItem In userDeclaration.References
                Set callSite = refItem
                Set callScope = callSite.ScopingDeclaration
                PrettyPrint ActiveSheet, userDeclaration, callScope, callSite.StartLine, callSite.StartColumn
            Next
        Next
End Sub

Private Sub PrettyPrint(ByVal target As Worksheet, ByVal userDeclaration As Declaration, ByVal callSite As IdentifierReference, ByVal callScope As Declaration, ByVal linePosition As Long, ByVal columnPosition As Long)

End Sub
````

That would actually give you all *used* declarations, including all variables; you can use `Declaration.DeclarationType` to filter and fine-tune the output - it could easily be tweaked to locate any unused procedures or variables.

I think we should find a way to *plot* the dependencies and dump it into a tearable tab in the editor (with a toolbar button to copy it to the clipboard), but that'll have to wait for 3.0 - in the meantime I think there's very little work involved in getting the call sites to show up in the *Code Explorer* itself, and there's already a "copy to clipboard" implementation for that toolwindow - I say we:

 - [ ] add identifier reference nodes to the code explorer
 - [ ] re-implement the CE's *export to clipboard* feature to copy some kind of *dependency map* of the project, a table showing who's called where...

Thoughts?

**Comment by Vogel612** [comment]
I'm not sure a table is the best option for this, since the caller-callee relation is a many-to-many, which makes it really hard to put this into a single table :/ Aside from that :+1:

**Comment by joshuader6** [comment]
@Vogel612  Yes i do agree on this point. But once you have it sorted in a Table you can do further analysis and sorting there. You can Filter alot and visualize it in some way.

**Comment by SystemsModelling** [comment]
Thanks for the example, retailcoder!

**Comment by IvenBach** [comment]
This could help build a map of what code is called where, something I could use myself.

**Comment by robodude666** [comment]
:+1:  It would definitely be useful to get a module-level and procedure-level dependency graph. It'll help identify unintentional dependencies.

Would be great to see a graph, but it's understandably a lot of work.

**Comment by Vogel612** [comment]
Related #4874

**Comment by Vogel612** [comment]
Reminder to self: You wanted to use https://www.nuget.org/packages/GraphX/ for this

**Comment by MDoerner** [comment]
Note that we already collect which module depends on which other module in the `ModuleToModuleReferenceManager`.

**Comment by bclothier** [comment]
We provide a way to find references via the toolbar. However, to make this even more useful, we could implement `FindReferencedModulesCommand` and a `FindReferencingModulesCommand`. We can then provide the ability to output via the toolwindow.
